src/hardware/serialport/softmodem.cpp

```diff
@@ -29,6 +29,75 @@
 #include "softmodem.h"
 #include "misc_util.h"
 
+static const char phoneValidChars[] = "01234567890*=,;#+>";
+
+class CPhonebookEntry {
+public:
+	CPhonebookEntry(const char *_phone, const char *_address) {
+		safe_strncpy(phone, _phone, sizeof(phone));
+		safe_strncpy(address, _address, sizeof(phone));
+	}
+
+	const char *IsMatchingPhone(const char *input) {
+		if (strcmp(input, phone) != 0)
+			return NULL;
+
+		return address;
+	}
+
+	static char IsValidPhone(const char *input) {
+		for (const char *ch = input; *ch; ch++) {
+			if (!strchr(phoneValidChars, *ch)) {
+				LOG_MSG("SERIAL: Phone %s contains invalid character %c", input, *ch);
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+private:
+	char phone[64];
+	char address[64];
+};
+
+static std::vector<CPhonebookEntry *> phones;
+
+bool MODEM_ReadPhonebook(const char *filename) {
+	LOG_MSG("SERIAL: Loading phonebook from %s", filename);
+
+	FILE *loadfile = fopen(filename, "rt");
+	if (!loadfile) return false;
+	char linein[512];
+	while (fgets(linein, 512, loadfile)) {
+		char *line = trim(linein);
+		char *phone = StripWord(line);
+		char *address = StripWord(line);
```

Comment:
std::string linein;
while (std::getline(loadfile, linein)) {
    std::istringstream iss(linein);
    std::string phone, address;
    if (!(iss >> phone >> address)) {
        // LOG_MSG about skipping the bad line
        continue;
    } 
}

Reply:
The above chunk relies on:
```
#include <sstream>
#include <string>
```

Reply:
If we stay with C, in theory the maximum line length is:

[whitespace] [max-phone-number-length] [whitespace] [max-hostname-length]:[max-port-number-length]

- International phone numbers max out at 15 digits, but people can insert `+`s and `-'s. So let's assume every number is prefixed with a character: so that doubles this to 30 characters.
- Hostnames max out at 255 characters per the RFC spec.  IPv6 IPs max out at 39 characters, so we're covered here too.
- Colon separator is one character
- The maximum internet port is 65535, or 5 characters (both IPv4 and 6)

So we're 30 + 255 + 1 + 5, or 291 characters.  I'm not sure how much whitespace is considered reasonable around these two values, but throw that in too. 

(refs)
https://www.ietf.org/rfc/rfc1035.txt
https://en.wikipedia.org/wiki/Telephone_numbering_plan#International_numbering_plan